add blog for feature ServiceNodePortStaticSubrange (#40204)
kubernetes/website: content/en/blog/_posts/2023-03-08-nodeport-dynamic-and-static-allocation.md

---
layout: blog
title: "Kubernetes 1.27: Avoid Collisions Assigning ports to Nodeport Services"
date: 2023-03-08
slug: nodeport-dynamic-and-static-allocation
---

> **Author:** xuzhenglun

Reviewer: Add your affiliation if possible.

Author: I need to confirm with the legal counsel whether it is possible to add

> In Kubernetes, the Service can be used to provide a unified traffic entry point for applications running on a set of Pods. Clients can use the virtual IP address provided by the Service for access, and Kubernetes provides load balancing for traffic accessing different back-end Pods, but ClusterIP-type Service is limited to providing access to nodes within the cluster, and traffic from outside the cluster cannot be routed. One way to solve this problem is to use NodePort-type Service, which maps a VIP's port to a specific port of all nodes in the cluster, thus redirecting traffic from the outside to the inside of the cluster.

Reviewer: Wrap long lines like this properly.
It will help reviewers to provide feedbacks;
and it will benefit downstream localization teams for translation and change tracking.

Author: It's done

---
layout: blog
title: "Kubernetes 1.27: Avoid Collisions Assigning ports to Nodeport Services"
date: 2023-03-08
slug: nodeport-dynamic-and-static-allocation
---

**Author:** xuzhenglun (Alibaba)

> In Kubernetes, the Service can be used to provide a unified traffic entry point for 

Reviewer: ```suggestion
In Kubernetes, a Service can be used to provide a unified traffic endpoint for
```

applications running on a set of Pods. Clients can use the virtual IP address provided
by the Service for access, and Kubernetes provides load balancing for traffic accessing

> different back-end Pods, but ClusterIP-type Service is limited to providing access to

Reviewer: ```suggestion
different back-end Pods, but a ClusterIP type of Service is limited to providing access to
```

> nodes within the cluster, and traffic from outside the cluster cannot be routed.

Reviewer: ```suggestion
nodes within the cluster, while traffic from outside the cluster cannot be routed.
```

> One way to solve this problem is to use NodePort-type Service, which maps a VIP's port

Reviewer: ```suggestion
One way to solve this problem is to use a NodePort type of Service, which maps a VIP's port
```

to a specific port of all nodes in the cluster, thus redirecting traffic from the
outside to the inside of the cluster.

## How NodePort's ports are allocated? 

> When a NodePort-type Service is created, its corresponding port is allocated in two ways.

Reviewer: ```suggestion
When a NodePort Service is created, its corresponding port is allocated in two ways.
```

**Dynamic** : If the Service type is NodePort and the corresponding `service.spec.ports.nodePort`

> is not specified, the Kubernetes controlpalne will automatically allocat a unused port 

Reviewer 2: ```suggestion
is not specified, the Kubernetes controlplane will automatically allocate an unused port 
```

Author: thanks, It's done

> **Dynamic** : If the Service type is NodePort and the corresponding `service.spec.ports.nodePort`
> is not specified, the Kubernetes controlplane will automatically allocate an unused port 
> to it at creation time.

Reviewer: ```suggestion
- **Dynamic** : If the Service type is NodePort and the corresponding `service.spec.ports.nodePort`
  is not specified, the Kubernetes control plane will automatically allocate an unused port 
  to it at creation time.
```

**Static** : In addition to the dynamic auto-assignment described above, you can also

> explicitly assign a port that is within the nodeport port configuration.

Reviewer 2: ```suggestion
explicitly assign a port that is within the nodeport port range configuration.
```

Author: thanks, It's done

> **Static** : In addition to the dynamic auto-assignment described above, you can also
> explicitly assign a port that is within the nodeport port range configuration.

Reviewer: ```suggestion
- **Static** :You can also explicitly assign a port that is within the NodePort port range configuration.
```

the `service.spec.ports.nodePort` must be unique for each NodePort-type Service across the cluster.

> Attempting to create a NodePort-type Service with a port already allocated will return an error.

Reviewer: ```suggestion
Attempting to create a NodePort Service with a port already allocated will return an error.
```

> ## Why do you need to reserve ports of NodePort-type Service? 

Reviewer: ```suggestion
## Why do you need to reserve ports for NodePort Services? 
```

> Sometimes, you may want to have the NodePort-type Service running on well-known ports

Reviewer: ```suggestion
Sometimes, you may want to have a NodePort Service running on well-known ports
```

so that other components and users inside and outside the cluster can use them.

In some complex cluster deployments with a mix of K8S nodes and non-K8S nodes, it may be

> necessary to rely on some pre-defined port for communicating. In particular, some fundamental

Reviewer: ```suggestion
necessary to use some pre-defined ports for communication. In particular, some fundamental
```

> components cannot rely on the VIPs which provided by LoadBalancer-type  services,

Reviewer: ```suggestion
components cannot rely on the VIP provided by a LoadBalancer Service,
```

> because the load balancing itself that provides the VIPs may also rely on these fundamental components.

Reviewer: ```suggestion
because the load balancer that provides the VIPs may also rely on these fundamental components.
```

> Now suppose we need to expose the Minio object storage service on K8S to services located on non-K8S nodes,

Reviewer: ```suggestion
Now suppose we need to expose a Minio object storage service on K8S to services located on non-K8S nodes,
```

and the agreed port is `30009`, we need to create a Service as follows:

```yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    k8s-app: minio
  name: minio
  namespace: kube-system
spec:
  ports:
  - name: http
    nodePort: 30009
    port: 9000
    protocol: TCP
    targetPort: 9000
  selector:
    app: minio
  type: NodePort
```

However, as mentioned before, if the port `30009` required for the minio Service is not reserved,

> and another NodePort Service is created and dynamically allocated before or concurrently with minio

Reviewer: ```suggestion
and another NodePort Service is created and dynamically allocated before or concurrently with the minio
```

> Service, `30009` can be allocated to those NodePort Service, and minio Service will fail to be

Reviewer: ```suggestion
Service, `30009` can be allocated to those NodePort Service, while the minio Service will fail to be
```

created due to a port conflict.

## How can you avoid NodePort-type Service port conflicts? 
In the Kubernetes 1.24, Service ClusterIP had already divided the CIDRs into two blocks, using different
allocation policies to reduce the risk of conflicts. In Kubernetes 1.27, a similar policy can be adopted

> for NodePort. You can enable a new featuregate `ServiceNodePortStaticSubrange`. Turning this on allows you

Reviewer: ```suggestion
for NodePort Services. You can enable a new feature gate `ServiceNodePortStaticSubrange`. Turning this on allows you
```

> to use a different port allocation strategy for NodePort-type Services, and reducing the risk of collision.

Reviewer: ```suggestion
to use a different port allocation strategy for NodePort Services, and reducing the risk of collision.
```

The port range for `NodePort` will be divided, based on the formula `min(max(16, nodeport-size / 32), 128)`,
which can be described as _never less than 16 or more than 128 with a graduated step between them_.

Dynamic port assignment will use the upper band by default, once this has been exhausted it will use the lower range.
This will allow users to use static allocations on the lower band with a low risk of collision.

## Examples:

### default range: 30000-32767
service-node-port-range: 30000-32767
Band Offset: `min(max(16, 2768/32), 128)` = `min(max(16, 86), 128)` = `min(86, 128)` = 86
Static band start: 30000
Static band end: 30085

> Range end: 32767

Reviewer: I don't think the output would be good for these lines.

Author: Would it be better to change them to tables?

---
layout: blog

> title: "Kubernetes 1.27: Avoid Collisions Assigning ports to Nodeport Services"

Reviewer 3: ```suggestion
title: "Kubernetes 1.27: Avoid Collisions Assigning Ports to NodePort Services"
```

Author: thanks, it's fixed

date: 2023-03-08
slug: nodeport-dynamic-and-static-allocation
---

> **Author:** xuzhenglun (Alibaba)

Reviewer 3: I'd expect to see something like:
```markdown
**Author:** Xu Zhenglun (Alibaba)
```
or
```markdown
**Author:** “Reficul” (Alibaba)
```

Author: done

In Kubernetes, a Service can be used to provide a unified traffic endpoint for 

> applications running on a set of Pods. Clients can use the virtual IP address provided

Reviewer 3: ```suggestion
applications running on a set of Pods. Clients can use the virtual IP address (or _VIP_) provided
```

Author: done

by the Service for access, and Kubernetes provides load balancing for traffic accessing
different back-end Pods, but a ClusterIP type of Service is limited to providing access to
nodes within the cluster, while traffic from outside the cluster cannot be routed.

> One way to solve this problem is to use a NodePort type of Service, which maps a VIP's port

Reviewer 3: ```suggestion
One way to solve this problem is to use a `type: NodePort` Service, which sets up a mapping
from the virtual IP address (the ClusterIP representing the Service)
```

Author: done

to a specific port of all nodes in the cluster, thus redirecting traffic from the
outside to the inside of the cluster.

> ## How NodePort's ports are allocated? 

Reviewer 3: ```suggestion
## How Kubernetes allocates node ports to Services
```
?

Author: it's fixed.

> When a NodePort Service is created, its corresponding port is allocated in two ways.

Reviewer 3: ```suggestion
When a `type: NodePort` Service is created, its corresponding port(s) are allocated in one
of two ways:
```

Author: done

> - **Dynamic** : If the Service type is NodePort and the corresponding `service.spec.ports.nodePort`
>   is not specified, the Kubernetes control plane will automatically allocate an unused port 
>   to it at creation time.

Reviewer 3: ```suggestion
- **Dynamic** : If the Service type is `NodePort` and you do not set a `nodePort` value
  explicitly in the `spec` for that Service, the Kubernetes control plane will automatically allocate
  an unused port to it at creation time.
```

- **Static** : In addition to the dynamic auto-assignment described above, you can also
  explicitly assign a port that is within the nodeport port range configuration.

> the `service.spec.ports.nodePort` must be unique for each NodePort Service across the cluster.

Reviewer 3: ```suggestion
The value of `nodePort` that you manually assign must be unique across the whole cluster.
```

> Attempting to create a NodePort type of Service with a port already allocated will return an error.

Reviewer 3: ```suggestion
Attempting to create a Service of `type: NodePort` where you explicitly specify a node port that
was already allocated results in an error.
```

## Why do you need to reserve ports of NodePort Service? 
Sometimes, you may want to have a NodePort Service running on well-known ports
so that other components and users inside o r outside the cluster can use them.

> In some complex cluster deployments with a mix of K8S nodes and non-K8S nodes, it may be

Reviewer 3: ```suggestion
In some complex cluster deployments with a mix of Kubernetes nodes and other servers on the same
network, it may be
```
If it's not running a kubelet, it's not a node. At least not on this website.

necessary to use some pre-defined ports for communication. In particular, some fundamental

> components cannot rely on the VIPs which provided by LoadBalancer services,

Reviewer 3: ```suggestion
components cannot rely on the VIPs that back `type: LoadBalancer` Services
```

> because the load balancer that provides the VIPs may also rely on these fundamental components.

Reviewer 3: ```suggestion
because the virtual IP address mapping implementation for that cluster also relies on
these foundational components.
```

Is this right? I am not sure I have understood the message intended here.

Author: done

> Now suppose we need to expose a Minio object storage service on K8S to services located on non-K8S nodes,

Reviewer 3: ```suggestion
Now suppose you need to expose a Minio object storage service on Kubernetes to clients running outside
the Kubernetes cluster,
```

> and the agreed port is `30009`, we need to create a Service as follows:

Reviewer 3: ```suggestion
and the agreed TCP port is 30009. You could create a Service as follows:
```
There are alternatives to using a Service here (though they aren't common).

```yaml
apiVersion: v1
kind: Service
metadata:

>   labels:
>     k8s-app: minio
>   name: minio
>   namespace: kube-system

Reviewer 3: ```suggestion
  name: minio
```
Don't suggest running Minio in `kube-system` becasue doing that is rarely considered good practice.

spec:
  ports:

>   - name: http

Reviewer 3: (nit)
It's slightly unidiomatic to set a port other than 80 if you name a port `http`.

Author: Well, the official minio chart names this port as `http` or `https`: https://github.com/minio/charts/blob/main/minio/templates/service.yaml#L1.

but I've changed it to its actual function `api` for clarity. Do you think that's okay for you?

Reviewer 3: This is a very small nit. Lots of people name ports “wrongly” against a strict reading of the Kubernetes API for Service.

    nodePort: 30009
    port: 9000
    protocol: TCP
    targetPort: 9000
  selector:
    app: minio
  type: NodePort
```

> However, as mentioned before, if the port `30009` required for the minio Service is not reserved,
> and another NodePort Service is created and dynamically allocated before or concurrently with the minio
> Service, `30009` can be allocated to those NodePort Service, while the minio Service will fail to be
> created due to a port conflict.

Reviewer 3: ```suggestion
However, as mentioned before, if the port (30009) required for the `minio` Service is not reserved,
and another `type: NodePort` (or possibly `type: LoadBalancer`) Service is created and dynamically
allocated before or concurrently with the `minio` Service, TCP port 30009 might be allocated to that
other Service; if so, creation of the `minio` Service will fail due to a node port collision.
```

## How can you avoid NodePort Service port conflicts? 

> In the Kubernetes 1.24, Service ClusterIP had already divided the CIDRs into two blocks, using different
> allocation policies to reduce the risk of conflicts. In Kubernetes 1.27, a similar policy can be adopted
> for NodePort Services. You can enable a new feature gate `ServiceNodePortStaticSubrange`. Turning this on allows you
> to use a different port allocation strategy for NodePort Services, and reducing the risk of collision.

Reviewer 3: ```suggestion
Kubernetes 1.24 introduced changes for `type: ClusterIP` Services, dividing the CIDR range for cluster
IP addresses into two blocks that use different allocation policies to [reduce the risk of conflicts](/docs/reference/networking/virtual-ips/#avoiding-collisions).
In Kubernetes 1.27, as an alpha feature, you can adopt a similar policy for `type: NodePort` Servies.
You can enable a new [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
`ServiceNodePortStaticSubrange`. Turning this on allows you to use a different port allocation strategy
for `type: NodePort` Services, and reduce the risk of collision.
```

The port range for `NodePort` will be divided, based on the formula `min(max(16, nodeport-size / 32), 128)`,

> which can be described as _never less than 16 or more than 128 with a graduated step between them_.

Reviewer 3: This doesn't make sense. What does the outcome of the formula determine?

If my cluster is running with the configuration `--service-node-port-range=65520-65535` and I enable this feature gate, what happens?

(“nobody would ever do that” doesn't feel like a realistic response, given that there's a similar example just a bit later)

Author: I modified some wording to explain the meaning of the formula output.

If `--service-node-port-range=65520-65535` is configured and the size of the node port range is set to 15, which is less than 16, then the size of the static port range should be set to 0. This means that all ports will be allocated using dynamic policy.

| Range properties        | Values                                                |
|-------------------------|-------------------------------------------------------|
| service-node-port-range | 30000-32767                                           |
| Band Offset             | &ensp; `min(max(16, 2768/32), 128)` <br>= `min(max(16, 86), 128)` <br>= `min(86, 128)` <br>= 86 |
| Static band start       | 30000                                                 |
| Static band end         | 30085                                                 |
| Range end               | 32767                                                 |

{{< mermaid >}}
pie showData
    title 30000-32767
    "Static" : 86
    "Dynamic" : 2682
{{< /mermaid >}}

### very small range: 30000-30015
| Range properties        | Values                                                |
|-------------------------|-------------------------------------------------------|
| service-node-port-range | 30000-30015                                           |
| Band Offset             | 0                                                     |
| Static band start       | 30000                                                 |

> | Static band end         | 30000                                                 |

Reviewer 3: Is port 30000 reserved for static assignment, then?

Author: no, port number 30000 is not reserved. I've added rows for 'dynamic band start/end' in the table, which should make it clearer.

---
layout: blog
title: "Kubernetes 1.27: Avoid Collisions Assigning Ports to NodePort Services"

> date: 2023-04-10

Reviewer: Please change this date and the filename according to the suggestion.

Author: thanks, it's done

In Kubernetes 1.27, as an alpha feature, you can adopt a similar policy for `type: NodePort` Servies.
You can enable a new [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
`ServiceNodePortStaticSubrange`. Turning this on allows you to use a different port allocation strategy
for `type: NodePort` Services, and reduce the risk of collision.

The port range for `NodePort` will be divided, based on the formula `min(max(16, nodeport-size / 32), 128)`. 
The outcome of the formula will be a number between 16 and 128, with a step size that increases as the 
size of the nodeport range increases. The outcome of the formula determine that the size of static port 
range. When the port range is less than 16, the size of static port range will be set to 0, 
which means that all ports will be dynamically allocated.

Dynamic port assignment will use the upper band by default, once this has been exhausted it will use the lower range.
This will allow users to use static allocations on the lower band with a low risk of collision.

> ## Examples:

Reviewer 3: (nit)
```suggestion
## Examples
```

Author: it's fixed.
